src/server/worker/api_server.go

```diff
@@ -957,3 +744,420 @@ func (a *APIServer) updateJobState(stm col.STM, jobInfo *pps.JobInfo, state pps.
 	jobs.Put(jobInfo.Job.ID, jobInfo)
 	return nil
 }
+
+type acquireDatumsFunc func(low, high int64) (failedDatumID string, _ error)
+
+func (a *APIServer) acquireDatums(ctx context.Context, jobID string, chunks *Chunks, logger *taggedLogger, process acquireDatumsFunc) error {
+	complete := false
+	for !complete {
+		var low int64
+		var high int64
+		var found bool
+		if _, err := col.NewSTM(ctx, a.etcdClient, func(stm col.STM) error {
+			found = false
+			locks := a.locks(jobID).ReadWrite(stm)
+			// we set complete to true and then unset it if we find an incomplete chunk
+			complete = true
+			for _, high = range chunks.Chunks {
+				var chunkState ChunkState
+				if err := locks.Get(fmt.Sprint(high), &chunkState); err != nil {
```

Comment:
Could we add `// Try to acquire a lease for chunk no. 'high'`?

Reply:
Also, just to confirm: this uses one lease for `k` documents, right? If so I'll update the design doc

Reply:
Sure thing on adding that comment.

And yes this is one lease for `k` documents.